### PR TITLE
Books ,Categories, Publishers and authors tables + add jQuery

### PR DESCRIPTION
**Pull request by fadyyousry**

### CODE REVIEW

File: app/javascript/packs/book/book.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        this.parentNode.appendChild(a);
+
+        var $title = $(this).val().replace(/\s+/g, '+');
+
+        $.ajax({
+            type: 'GET',
+            url: 'https://www.googleapis.com/books/v1/volumes',
+            data: { "q": $title, "maxResults": "3" },
+            success: function(books) {
+                var list = $("#search_google_book_list");
```
Comment (fadyyousry) [inline app/javascript/packs/book/book.js:29]: remove

File: Gemfile [hunk 2]
```diff
@@ -1,59 +0,0 @@
...
-gem 'rails', '~> 6.0.3', '>= 6.0.3.2'
-# Use sqlite3 as the database for Active Record
-gem 'sqlite3', '~> 1.4'
-# Use Puma as the app server
-gem 'puma', '~> 4.1'
-# Use SCSS for stylesheets
-gem 'sass-rails', '>= 6'
-# Transpile app-like JavaScript. Read more: https://github.com/rails/webpacker
-gem 'webpacker', '~> 4.0'
-# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
```
Comment (Magdyx) [inline Gemfile:16]: Gemfile shouldn't be deleted. Please fix this.

File: app/assets/stylesheets/categories.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the categories controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (Magdyx) [inline app/assets/stylesheets/categories.scss:3]: please don't commit empty files.

File: app/assets/stylesheets/google_books.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the google_books controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (Magdyx) [inline app/assets/stylesheets/google_books.scss:3]: ditto

File: app/assets/stylesheets/publishers.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the publishers controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (Magdyx) [inline app/assets/stylesheets/publishers.scss:3]: ditto

File: app/assets/stylesheets/scaffolds.scss [hunk 6]
```diff
@@ -0,0 +1,65 @@
+body {
```
Comment (Magdyx) [inline app/assets/stylesheets/scaffolds.scss:1]: name of the file is not self-explanatory

File: app/assets/stylesheets/scaffolds.scss [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  background-color: #fff;
+  color: #333;
+  margin: 33px; }
```
Comment (Magdyx) [inline app/assets/stylesheets/scaffolds.scss:4]: usually don't mess with the body as it affects the whole project

File: app/assets/stylesheets/scaffolds.scss [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  font-family: verdana, arial, helvetica, sans-serif;
+  font-size: 13px;
+  line-height: 18px; }
+
+pre {
+  background-color: #eee;
+  padding: 10px;
+  font-size: 11px; }
+
+a {
```
Comment (Magdyx) [inline app/assets/stylesheets/scaffolds.scss:16]: also don't mess with tags like that, use classes and ids

File: app/assets/stylesheets/scaffolds.scss [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  font-size: 12px;
+  margin: -7px -7px 0;
+  background-color: #c00;
+  color: #fff; }
+
+#error_explanation ul li {
+  font-size: 12px;
+  list-style: square; }
+
+label {
```
Comment (Magdyx) [inline app/assets/stylesheets/scaffolds.scss:64]: ditto

File: app/controllers/authors_controller.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  def destroy
+    @author.destroy
+    respond_to do |format|
+      format.html { redirect_to authors_url, notice: 'Author was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
```
Comment (Magdyx) [inline app/controllers/authors_controller.rb:65]: having comments is a bad practice in ruby.
avoid intuitive comments like this one.
Most comments should be understood from reading the code and using proper naming conventions. 
If you feel the urge to convey an important message add it in the commit message.

File: app/controllers/authors_controller.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
+class AuthorsController < ApplicationController
```
Comment (Magdyx) [inline app/controllers/authors_controller.rb:1]: I don't think there should be an `AuthorsController`.
If you don't use it please remove it.

File: app/controllers/authors_controller.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  def create
+    @author = Author.new(author_params)
+
+    respond_to do |format|
+      if @author.save
+        format.html { redirect_to @author, notice: 'Author was successfully created.' }
+        format.json { render :show, status: :created, location: @author }
+      else
+        format.html { render :new }
+        format.json { render json: @author.errors, status: :unprocessable_entity }
```
Comment (Magdyx) [inline app/controllers/authors_controller.rb:35]: what are you using json response for?

File: app/controllers/authors_controller.rb [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      end
+    end
+  end
+
+  # DELETE /authors/1
+  # DELETE /authors/1.json
+  def destroy
+    @author.destroy
+    respond_to do |format|
+      format.html { redirect_to authors_url, notice: 'Author was successfully destroyed.' }
```
Comment (Magdyx) [inline app/controllers/authors_controller.rb:59]: I think this kind of message needs to be localized

File: app/controllers/books_controller.rb [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  # POST /books
+  # POST /books.json
+  def create
+    @book = Book.new(book_params)
+
+    respond_to do |format|
+      if @book.save
+        format.html { redirect_to @book, notice: 'Book was successfully created.' }
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:31]: ditto

File: app/controllers/books_controller.rb [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        format.json { render :show, status: :created, location: @book }
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:32]: ditto

File: app/controllers/books_controller.rb [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+      else
+        format.html { render :new }
+        format.json { render json: @book.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH/PUT /books/1
+  # PATCH/PUT /books/1.json
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:41]: please remove this type of comments

File: app/controllers/books_controller.rb [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+  def update
+    respond_to do |format|
+      if @book.update(book_params)
+        format.html { redirect_to @book, notice: 'Book was successfully updated.' }
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:45]: ditto

File: app/controllers/books_controller.rb [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        format.json { render :show, status: :ok, location: @book }
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:46]: ditto

File: app/controllers/books_controller.rb [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+      end
+    end
+  end
+
+  # DELETE /books/1
+  # DELETE /books/1.json
+  def destroy
+    @book.destroy
+    respond_to do |format|
+      format.html { redirect_to books_url, notice: 'Book was successfully destroyed.' }
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:59]: ditto

File: app/controllers/books_controller.rb [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_book
+      @book = Book.find(params[:id])
+    end
+
+    # Only allow a list of trusted parameters through.
+    def book_params
+      params.require(:book).permit(:title, :publisher_id, :published_date, :description, :isbn, :page_count, :image_link, :language, :is_pdf, :quantity,:price)
```
Comment (Magdyx) [inline app/controllers/books_controller.rb:72]: usually break long lines for better readability, 
Also insert space before `:price`

File: app/controllers/categories_controller.rb [hunk 21]
```diff
@@ -0,0 +1,74 @@
+class CategoriesController < ApplicationController
```
Comment (Magdyx) [inline app/controllers/categories_controller.rb:1]: I don't think we need this controller either

File: app/controllers/publishers_controller.rb [hunk 22]
```diff
@@ -0,0 +1,74 @@
+class PublishersController < ApplicationController
```
Comment (Magdyx) [inline app/controllers/publishers_controller.rb:1]: same as categories and publishers

File: app/helpers/authors_helper.rb [hunk 23]
```diff
@@ -0,0 +1,2 @@
+module AuthorsHelper
```
Comment (Magdyx) [inline app/helpers/authors_helper.rb:1]: also remove those

File: app/helpers/books_helper.rb [hunk 24]
```diff
@@ -0,0 +1,2 @@
+module BooksHelper
```
Comment (Magdyx) [inline app/helpers/books_helper.rb:1]: please remove

File: app/helpers/categories_helper.rb [hunk 25]
```diff
@@ -0,0 +1,2 @@
+module CategoriesHelper
```
Comment (Magdyx) [inline app/helpers/categories_helper.rb:1]: same

File: app/helpers/google_books_helper.rb [hunk 26]
```diff
@@ -0,0 +1,2 @@
+module GoogleBooksHelper
```
Comment (Magdyx) [inline app/helpers/google_books_helper.rb:1]: same

File: app/helpers/publishers_helper.rb [hunk 27]
```diff
@@ -0,0 +1,2 @@
+module PublishersHelper
```
Comment (Magdyx) [inline app/helpers/publishers_helper.rb:1]: same

File: app/javascript/packs/application.js [hunk 28]
```diff
@@ -7,11 +7,12 @@ require("@rails/ujs").start()
 require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
-
+require('jquery')
+require('packs/book/book.js')
```
Comment (Magdyx) [inline app/javascript/packs/application.js:11]: does it have to specify the entire path?
will `require('packs')` work?

File: app/javascript/packs/book/book.js [hunk 29]
```diff
@@ -0,0 +1,62 @@
+$(function() {
+    var $mybook;
+    /*execute a function when someone clicks in the document:*/
+    document.addEventListener("click", function(e) {
+        $(".autocomplete-items").remove();
+    });
+
+    $('#search_google_book_input').on('input', function() {
+        var inp = this;
+        var a, b, val = this.value;
```
Comment (Magdyx) [inline app/javascript/packs/book/book.js:10]: please use proper variables names

File: app/javascript/packs/book/book.js [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+                    b.innerHTML += "<input type='hidden' value='" + i + "'>";
+                    /*execute a function when someone clicks on the item value (DIV element):*/
+                    b.addEventListener("click", function(e) {
+                        /*insert the value for the autocomplete text field:*/
+                        inp.value = this.getElementsByTagName("p")[0].outerText;
+                        $mybook = books.items[this.getElementsByTagName('input')[0].value];
+                        /*close the list of autocompleted values,
+                        (or any other open lists of autocompleted values:*/
+                        $(".autocomplete-items").remove();
+                    });
```
Comment (Magdyx) [inline app/javascript/packs/book/book.js:46]: please remove intuitive comments

File: app/views/authors/_author.json.jbuilder [hunk 31]
```diff
@@ -0,0 +1,2 @@
+json.extract! author, :id, :name, :created_at, :updated_at
```
Comment (Magdyx) [inline app/views/authors/_author.json.jbuilder:1]: remove

File: config/routes.rb [hunk 32]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :categories
+  resources :books
+  resources :publishers
+  resources :authors
```
Comment (Magdyx) [inline config/routes.rb:5]: remove

**Pull request merged**